browser/ui/webui/tab_search/tab_search_page_handler_browsertest.cc

```diff
@@ -51,14 +66,65 @@ constexpr char kFooDotComTitle2[] = "foo.com 2";
 constexpr char kBarDotComTitle1[] = "bar.com 1";
 constexpr char kBarDotComTitle2[] = "bar.com 2";
 
+constexpr char kTopic[] = "topic";
+constexpr char kTopic2[] = "topic2";
+
 }  // namespace
 
 using testing::_;
 
+class TabChangeWaiter : ai_chat::mojom::TabDataObserver {
+ public:
+  explicit TabChangeWaiter(content::BrowserContext* context) {
+    auto* tracker_service =
+        ai_chat::TabTrackerServiceFactory::GetInstance()->GetForBrowserContext(
+            context);
+    tracker_service->AddObserver(receiver_.BindNewPipeAndPassRemote());
+  }
+
+  ~TabChangeWaiter() override = default;
+
+  void WaitForTabDataChanged(const std::vector<std::string>& titles) {
+    expected_titles_ = titles;
+    base::RunLoop run_loop;
+    quit_closure_ = run_loop.QuitClosure();
+    run_loop.Run();
+  }
+
+  void TabDataChanged(
+      std::vector<ai_chat::mojom::TabDataPtr> tab_data) override {
+    std::vector<std::string> titles;
+    for (const auto& tab : tab_data) {
+      titles.push_back(tab->title);
+    }
+
+    if (!quit_closure_ || titles != expected_titles_) {
```

Comment:
oh interesting - I didn't realise std::vector did an element-wise compare for equality!